### PR TITLE
Add endpoint to return parsed fda submissions

### PR DESCRIPTION
**Pull request by calvinlu3**
This is related to https://github.com/oncokb/oncokb/issues/3057

- For `FdaSubmission` entities, the `Number, Device Name, Companion Diagnostic Device, and Fda Submission Type ` are required
   - Also, cannot add new entity where `(number + supplement + cdx)` already exists
- Add endpoint at `/api/fda-submissions/lookup` to find an fda submission. If it doesn't already exist in db, then we use the parse the information from fda website
- Created `OncoKBTable.tsx`, which is a simple table using `react-table v7`
- Created our own [ValidatedForm](https://github.com/calvinlu3/oncokb-transcript/blob/fda-submission-endpoint/src/main/webapp/app/shared/form/ValidatedForm.tsx) and [ValidatedField](https://github.com/calvinlu3/oncokb-transcript/blob/fda-submission-endpoint/src/main/webapp/app/shared/form/ValidatedField.tsx), so form elements can be nested using `react-hook-forms`

### CODE REVIEW

File: src/main/webapp/app/entities/companion-diagnostic-device/companion-diagnostic-device.tsx [hunk 1]
```diff
@@ -25,6 +27,35 @@ export const CompanionDiagnosticDevice = (props: ICompanionDiagnosticDeviceProps
 
   const { match } = props;
 
+  const columns: Column<ICompanionDiagnosticDevice>[] = [
+    { id: 'id', accessor: 'id', Header: 'ID' },
```
Comment (zhx828) [inline src/main/webapp/app/entities/companion-diagnostic-device/companion-diagnostic-device.tsx:31]: Do we need to show the id if we don't allow clicking through the id to the device page like we used to do?

File: src/main/webapp/app/entities/fda-submission/fda-submission.store.ts [hunk 2]
```diff
@@ -1,14 +1,33 @@
...
+    makeObservable(this, {
+      lookupFdaSubmission: action,
+    });
+  }
+
+  *lookupFdaSubmissionGen(submissionNumber: string) {
+    const numbers = submissionNumber.split('/');
+    const url = `${apiUrl}/lookup?number=${numbers[0]}${numbers[1] ? `&supplement_number=${numbers[1]}` : ''}`;
+    try {
+      const result: AxiosResponse<IFdaSubmission> = yield axios.get(url);
```
Comment (zhx828) [inline src/main/webapp/app/entities/fda-submission/fda-submission.store.ts:24]: Why do we use `yield` here?

File: src/main/webapp/app/entities/fda-submission/fda-submission.store.ts [hunk 3]
```diff
@@ -1,14 +1,33 @@
...
+      this.reset();
+      this.entity = result.data ? result.data : {};
```
Comment (zhx828) [inline src/main/webapp/app/entities/fda-submission/fda-submission.store.ts:26]:
```suggestion
      this.entity = result.data || {};
```

File: src/main/webapp/app/entities/fda-submission/fda-submission.tsx [hunk 4]
```diff
@@ -77,6 +79,63 @@ export const FdaSubmission = (props: IFdaSubmissionProps) => {
 
   const { match } = props;
 
+  const columns: Column<IFdaSubmission>[] = [
+    {
+      id: 'id',
+      accessor: 'id',
```
Comment (zhx828) [inline src/main/webapp/app/entities/fda-submission/fda-submission.tsx:85]: I remember if the accessor is not specified, it uses id. Could you take a look? Though I do like list accessor always to make it clear.

File: src/main/webapp/app/shared/form/ValidatedForm.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import React, { useEffect } from 'react';
+import { DefaultValues, FieldValues, FormProvider, SubmitHandler, useForm, useFormContext, ValidationMode } from 'react-hook-form';
+import { Form } from 'reactstrap';
+
+interface IValidatedFormProps {
+  children: React.ReactNode;
```
Comment (zhx828) [inline src/main/webapp/app/shared/form/ValidatedForm.tsx:6]: You do not need to specify children.

**Review by zhx828 — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/entities/companion-diagnostic-device/companion-diagnostic-device.tsx (hunk 1 above)
Comment (calvinlu3) [reply src/main/webapp/app/entities/companion-diagnostic-device/companion-diagnostic-device.tsx:31]: We can omit that information. Also, just realized that for react-table columns, we don't need to `id` field if `accessor` is a string.

File: src/main/webapp/app/entities/fda-submission/fda-submission.tsx (hunk 4 above)
Comment (calvinlu3) [reply src/main/webapp/app/entities/fda-submission/fda-submission.tsx:85]: If the accessor is a string (not function), then `id` is not necessary. I will remove the `id` field.

File: src/main/webapp/app/entities/fda-submission/fda-submission.store.ts (hunk 2 above)
Comment (calvinlu3) [reply src/main/webapp/app/entities/fda-submission/fda-submission.store.ts:24]: `generator-jhipster-react-mobx` uses [mobx flows](https://mobx.js.org/actions.html#asynchronous-actions), which takes in a generator function, for handling async methods.

File: src/main/webapp/app/config/constants.ts [hunk 6]
```diff
@@ -32,3 +32,5 @@ export enum PAGE_ROUTE {
   OAUTH = '/oauth2/authorization/oidc',
   ADMIN_USER_MANAGEMENT = '/admin/user-management',
 }
+
+export const FDA_SUBMISSION_REGEX = new RegExp('^([A-Z]+[0-9]{6})(\\/((S[0-9]{3})(-(S[0-9]{3}))?))?');
```
[the remark this answers is not in the document]
Comment (calvinlu3) [reply src/main/webapp/app/config/constants.ts:36]: We will remove this restriction.

File: src/main/java/org/mskcc/oncokb/curation/service/FdaSubmissionService.java [hunk 7]
```diff
@@ -99,6 +105,57 @@ public Optional<FdaSubmission> findOne(Long id) {
...
+     * @return Optional with the parsed fda submission or the one already existing in db, otherwise empty optional
+     */
+    public Optional<FdaSubmission> findOrFetchFdaSubmissionByNumber(String number, String supplementNumber) {
+        Optional<FdaSubmission> fdaSubmission = this.findByNumberAndSupplementNumber(number, supplementNumber);
+        if (fdaSubmission.isEmpty()) { // Fetch from FDA website if not present in database
+            if (StringUtils.isEmpty(number)) {
+                return Optional.empty();
+            }
+            String submissionNumber = number;
+            if (StringUtils.isEmpty(supplementNumber)) {
```
Comment (zhx828) [inline src/main/java/org/mskcc/oncokb/curation/service/FdaSubmissionService.java:131]:
```suggestion
            if (!StringUtils.isEmpty(supplementNumber)) {
```

**Pull request merged**